pallets/xcmp-queue/src/lib.rs

```diff
@@ -941,84 +510,160 @@ impl<T: Config> Pallet<T> {
 					s[index].state = OutboundState::Ok;
 				}
 			} else {
-				debug_assert!(false, "WARNING: Attempt to resume channel that was not suspended.");
+				defensive!("WARNING: Attempt to resume channel that was not suspended.");
 			}
 		});
 	}
+
+	fn enqueue_xcmp_messages(
+		sender: ParaId,
+		mut xcms: Vec<BoundedVec<u8, MaxXcmpMessageLenOf<T>>>,
+		meter: &mut WeightMeter,
+	) {
+		if meter
+			.try_consume(T::WeightInfo::enqueue_xcmp_messages(xcms.len() as u32))
+			.is_err()
+		{
+			return
+		}
+		let QueueConfigData { drop_threshold, .. } = <QueueConfig<T>>::get();
+		let fp = T::XcmpQueue::footprint(sender);
+
+		let new_count = xcms.len().saturating_add(fp.count as usize);
+		let to_enqueue = (drop_threshold as usize).saturating_sub(new_count) as usize;
+		if to_enqueue < xcms.len() {
+			// This should not happen since the channel should have been suspended in [`on_queue_changed`].
+			log::error!(
+				"XCMP queue for sibling {:?} is full; dropping {} messages.",
+				sender,
+				xcms.len() - to_enqueue,
+			);
+			xcms.truncate(to_enqueue);
+		}
+
+		T::XcmpQueue::enqueue_messages(xcms.iter().map(|xcm| xcm.as_bounded_slice()), sender);
+	}
+
+	/// Split concatenated encoded `VersionedXcm`s into individual items.
+	///
+	/// We directly encode them again since that is needed later on.
+	fn split_concatenated_xcms(
+		data: &mut &[u8],
+	) -> Result<Vec<BoundedVec<u8, MaxXcmpMessageLenOf<T>>>, ()> {
+		// FAIL-CI add benchmark to check for OOM depending on `MAX_XCM_DECODE_DEPTH`.
+		let mut encoded_xcms = Vec::new();
+		while !data.is_empty() {
+			let xcm =
+				VersionedXcm::<T::RuntimeCall>::decode_with_depth_limit(MAX_XCM_DECODE_DEPTH, data)
+					.map_err(|_| ())?;
+			let bounded = xcm.encode().try_into().map_err(|_| ())?;
+			encoded_xcms.push(bounded);
+		}
+		Ok(encoded_xcms)
+	}
+}
+
+impl<T: Config> OnQueueChanged<ParaId> for Pallet<T> {
```

Comment:
I guess it's probably coming in next commits, but just wanted to mention that it should probably be changed to:
```rust
impl<T: Config> OnQueueChanged<cumulus_primitives_core::AggregateMessageOrigin> for Pallet<T> {
```

and then used in `pallet_message_queue` configuration at all runtimes:
```rust
type QueueChangeHandler = XcmpQueue;
```

Now the `QueueChangeHandler` is set to `()` everywhere, which means that backpressure mechanism will never be activated.

Reply:
Ouch 🤦‍♂️ good catch!  

The `XcmpQueue` pallet is only ever able to handle `ParaId`s , no the full `AggregateMessageOrigin` so I am re-using the narrowing approach and now set:  
`type QueueChangeHandler = NarrowOriginToSibling<XcmpQueue>;`

Reply:
Ok - both options LGTM :) Thanks!